File: 399. Evaluate Division.java
class Solution {
    class Pair{
        String vertex;
        double weight;
        public Pair(String vertex, double weight){
            this.vertex = vertex;
            this.weight = weight;
        }
    }
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        HashMap<String, ArrayList<Pair>> adj = new HashMap<>();
        for(int i = 0; i < values.length; i++){
            ArrayList<Pair> original = adj.getOrDefault(equations.get(i).get(0), new ArrayList<>());
            ArrayList<Pair> reverse = adj.getOrDefault(equations.get(i).get(1), new ArrayList<>());
        
            original.add(new Pair(equations.get(i).get(1), values[i]));
            reverse.add(new Pair(equations.get(i).get(0), 1.0/values[i]));
            adj.put(equations.get(i).get(0), original);
            adj.put(equations.get(i).get(1), reverse);
        }


        double ans[] = new double[queries.size()];
        int index = 0;
        for(List<String> query : queries){
            HashMap<String, Double> visited = new HashMap<>();

            Queue<Pair> q = new LinkedList<>();
            if(!adj.containsKey(query.get(0))){
                ans[index++] = -1;
                continue;
            }
            visited.put(query.get(0), 1.0);
            q.offer(new Pair(query.get(0), 1.0));
            while(!q.isEmpty()){
                Pair top = q.poll();
                // System.out.println(top.vertex + "-" + top.weight);
                if(!adj.containsKey(top.vertex)) continue;
                
                for(Pair neig : adj.get(top.vertex)){
                    // System.out.println(neig.vertex + "-" + neig.weight);
                    if(!visited.containsKey(neig.vertex)){
                        q.offer(new Pair(neig.vertex, neig.weight * top.weight));
                        visited.put(neig.vertex, top.weight * neig.weight);
                    }
                    // else if(visited.get(neig.vertex) != top.weight * neig.weight) visited.put(neig.vertex, -1.0);
                }

                // System.out.println("----");
            }
            ans[index++] = visited.containsKey(query.get(1)) ? visited.get(query.get(1)) : -1;
        }

        return ans;
    }
}

File: 399. Evaluate Division.cpp
class Solution {
    struct Pair {
        string vertex;
        double weight;
        Pair(const string& vertex, double weight) : vertex(vertex), weight(weight) {}
    };
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        unordered_map<string, vector<Pair>> adj;
        for (size_t i = 0; i < values.size(); i++) {
            adj[equations[i][0]].emplace_back(equations[i][1], values[i]);
            adj[equations[i][1]].emplace_back(equations[i][0], 1.0 / values[i]);
        }

        vector<double> ans(queries.size());
        int index = 0;
        for (const auto& query : queries) {
            unordered_map<string, double> visited;
            queue<Pair> q;

            if (adj.find(query[0]) == adj.end()) {
                ans[index++] = -1;
                continue;
            }

            visited[query[0]] = 1.0;
            q.emplace(query[0], 1.0);

            while (!q.empty()) {
                Pair top = q.front();
                q.pop();

                if (adj.find(top.vertex) == adj.end()) continue;

                for (const auto& neig : adj[top.vertex]) {
                    if (visited.find(neig.vertex) == visited.end()) {
                        double new_weight = top.weight * neig.weight;
                        q.emplace(neig.vertex, new_weight);
                        visited[neig.vertex] = new_weight;
                    }
                }
            }

            ans[index++] = visited.count(query[1]) ? visited[query[1]] : -1;
        }

        return ans;
    }
};